File: api/src/main/java/com/ruby/rubia_server/core/controller/UserAIAgentController.java
package com.ruby.rubia_server.core.controller;

import com.ruby.rubia_server.core.base.BaseCompanyEntityController;
import com.ruby.rubia_server.core.dto.UserAIAgentDTO;
import com.ruby.rubia_server.core.dto.CreateUserAIAgentDTO;
import com.ruby.rubia_server.core.dto.UpdateUserAIAgentDTO;
import com.ruby.rubia_server.core.entity.UserAIAgent;
import com.ruby.rubia_server.core.service.UserAIAgentService;
import com.ruby.rubia_server.core.util.CompanyContextUtil;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/user-ai-agents")
@Slf4j
public class UserAIAgentController extends BaseCompanyEntityController<UserAIAgent, CreateUserAIAgentDTO, UpdateUserAIAgentDTO, UserAIAgentDTO> {

    private final UserAIAgentService userAIAgentService;

    public UserAIAgentController(UserAIAgentService userAIAgentService, CompanyContextUtil companyContextUtil) {
        super(userAIAgentService, companyContextUtil);
        this.userAIAgentService = userAIAgentService;
    }

    @Override
    protected String getEntityName() {
        return "UserAIAgent";
    }

    @Override
    protected UserAIAgentDTO convertToDTO(UserAIAgent userAIAgent) {
        return UserAIAgentDTO.builder()
                .id(userAIAgent.getId())
                .companyId(userAIAgent.getCompany().getId())
                .companyName(userAIAgent.getCompany().getName())
                .userId(userAIAgent.getUser().getId())
                .userName(userAIAgent.getUser().getName())
                .aiAgentId(userAIAgent.getAiAgent().getId())
                .aiAgentName(userAIAgent.getAiAgent().getName())
                .isDefault(userAIAgent.getIsDefault())
                .assignedAt(userAIAgent.getAssignedAt())
                .createdAt(userAIAgent.getCreatedAt())
                .updatedAt(userAIAgent.getUpdatedAt())
                .build();
    }

    @Override
    protected UUID getCompanyIdFromDTO(CreateUserAIAgentDTO createDTO) {
        return createDTO.getCompanyId();
    }

    // Endpoints específicos da entidade
    @GetMapping("/user/{userId}")
    public ResponseEntity<List<UserAIAgentDTO>> findByUserId(@PathVariable UUID userId) {
        log.debug("Finding UserAIAgents by user id via API: {}", userId);
        
        List<UserAIAgent> userAIAgents = userAIAgentService.findByUserId(userId);
        List<UserAIAgentDTO> responseDTOs = userAIAgents.stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
        
        return ResponseEntity.ok(responseDTOs);
    }

    @GetMapping("/ai-agent/{aiAgentId}")
    public ResponseEntity<List<UserAIAgentDTO>> findByAiAgentId(@PathVariable UUID aiAgentId) {
        log.debug("Finding UserAIAgents by ai agent id via API: {}", aiAgentId);
        
        List<UserAIAgent> userAIAgents = userAIAgentService.findByAiAgentId(aiAgentId);
        List<UserAIAgentDTO> responseDTOs = userAIAgents.stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
        
        return ResponseEntity.ok(responseDTOs);
    }

    @GetMapping("/user/{userId}/ai-agent/{aiAgentId}")
    public ResponseEntity<UserAIAgentDTO> findByUserIdAndAiAgentId(
            @PathVariable UUID userId, 
            @PathVariable UUID aiAgentId) {
        log.debug("Finding UserAIAgent by user id: {} and ai agent id: {}", userId, aiAgentId);
        
        Optional<UserAIAgent> userAIAgent = userAIAgentService.findByUserIdAndAiAgentId(userId, aiAgentId);
        if (userAIAgent.isEmpty()) {
            return ResponseEntity.notFound().build();
        }
        
        UserAIAgentDTO responseDTO = convertToDTO(userAIAgent.get());
        return ResponseEntity.ok(responseDTO);
    }

    @GetMapping("/default/{isDefault}")
    public ResponseEntity<List<UserAIAgentDTO>> findByIsDefault(@PathVariable Boolean isDefault) {
        log.debug("Finding UserAIAgents by isDefault via API: {}", isDefault);
        
        List<UserAIAgent> userAIAgents = userAIAgentService.findByIsDefault(isDefault);
        List<UserAIAgentDTO> responseDTOs = userAIAgents.stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
        
        return ResponseEntity.ok(responseDTOs);
    }

    @GetMapping("/user/{userId}/default")
    public ResponseEntity<UserAIAgentDTO> findByUserIdAndIsDefault(@PathVariable UUID userId) {
        log.debug("Finding default UserAIAgent for user id via API: {}", userId);
        
        Optional<UserAIAgent> userAIAgent = userAIAgentService.findByUserIdAndIsDefault(userId, true);
        if (userAIAgent.isEmpty()) {
            return ResponseEntity.notFound().build();
        }
        
        UserAIAgentDTO responseDTO = convertToDTO(userAIAgent.get());
        return ResponseEntity.ok(responseDTO);
    }

    @GetMapping("/user/{userId}/ai-agent/{aiAgentId}/exists")
    public ResponseEntity<Boolean> existsByUserIdAndAiAgentId(
            @PathVariable UUID userId, 
            @PathVariable UUID aiAgentId) {
        log.debug("Checking if UserAIAgent exists by user id: {} and ai agent id: {}", userId, aiAgentId);
        
        boolean exists = userAIAgentService.existsByUserIdAndAiAgentId(userId, aiAgentId);
        return ResponseEntity.ok(exists);
    }

    @GetMapping("/count/ai-agent/{aiAgentId}")
    public ResponseEntity<Long> countByAiAgentId(@PathVariable UUID aiAgentId) {
        log.debug("Counting UserAIAgents by ai agent id via API: {}", aiAgentId);
        
        long count = userAIAgentService.countByAiAgentId(aiAgentId);
        return ResponseEntity.ok(count);
    }

    @GetMapping("/count/user/{userId}")
    public ResponseEntity<Long> countByUserId(@PathVariable UUID userId) {
        log.debug("Counting UserAIAgents by user id via API: {}", userId);
        
        long count = userAIAgentService.countByUserId(userId);
        return ResponseEntity.ok(count);
    }

    @GetMapping("/count/default/{isDefault}")
    public ResponseEntity<Long> countByIsDefault(@PathVariable Boolean isDefault) {
        log.debug("Counting UserAIAgents by isDefault via API: {}", isDefault);
        
        long count = userAIAgentService.countByIsDefault(isDefault);
        return ResponseEntity.ok(count);
    }

    @PutMapping("/{id}/set-default")
    public ResponseEntity<UserAIAgentDTO> setAsDefault(
            @PathVariable UUID id, 
            @RequestParam Boolean isDefault) {
        log.debug("Setting UserAIAgent as default via API with id: {} and status: {}", id, isDefault);
        
        Optional<UserAIAgent> updated = userAIAgentService.setAsDefault(id, isDefault);
        if (updated.isEmpty()) {
            return ResponseEntity.notFound().build();
        }
        
        UserAIAgentDTO responseDTO = convertToDTO(updated.get());
        return ResponseEntity.ok(responseDTO);
    }

    @DeleteMapping("/user/{userId}/clear-defaults")
    public ResponseEntity<Void> clearDefaultForUser(@PathVariable UUID userId) {
        log.debug("Clearing default UserAIAgent for user via API: {}", userId);
        
        userAIAgentService.clearDefaultForUser(userId);
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/assign")
    public ResponseEntity<UserAIAgentDTO> assignUserToAgent(
            @RequestParam UUID userId,
            @RequestParam UUID aiAgentId,
            @RequestParam(defaultValue = "false") Boolean isDefault) {
        log.debug("Assigning user: {} to AI agent: {} with default: {}", userId, aiAgentId, isDefault);
        
        UserAIAgent userAIAgent = userAIAgentService.assignUserToAgent(userId, aiAgentId, isDefault);
        UserAIAgentDTO responseDTO = convertToDTO(userAIAgent);
        
        return ResponseEntity.ok(responseDTO);
    }

    @DeleteMapping("/user/{userId}/ai-agent/{aiAgentId}")
    public ResponseEntity<Void> removeUserFromAgent(
            @PathVariable UUID userId, 
            @PathVariable UUID aiAgentId) {
        log.debug("Removing user: {} from AI agent: {}", userId, aiAgentId);
        
        boolean removed = userAIAgentService.removeUserFromAgent(userId, aiAgentId);
        if (!removed) {
            return ResponseEntity.notFound().build();
        }
        
        return ResponseEntity.noContent().build();
    }
}

File: api-dotnet/Rubia.Server/Controllers/UserAIAgentController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Rubia.Server.DTOs;
using Rubia.Server.Entities;
using Rubia.Server.Services;

namespace Rubia.Server.Controllers;

[ApiController]
[Route("api/user-ai-agents")]
[Authorize]
public class UserAIAgentController : ControllerBase
{
    private readonly UserAIAgentService _userAIAgentService;
    private readonly CompanyContextService _companyContextService;
    private readonly ILogger<UserAIAgentController> _logger;

    public UserAIAgentController(
        UserAIAgentService userAIAgentService,
        CompanyContextService companyContextService,
        ILogger<UserAIAgentController> logger)
    {
        _userAIAgentService = userAIAgentService;
        _companyContextService = companyContextService;
        _logger = logger;
    }

    [HttpPost]
    public async Task<ActionResult<UserAIAgentDto>> Create([FromBody] CreateUserAIAgentDto createDto)
    {
        _logger.LogDebug("Creating UserAIAgent assignment: User {UserId} to AI Agent {AiAgentId}",
            createDto.UserId, createDto.AiAgentId);

        var userAIAgent = await _userAIAgentService.CreateAsync(
            createDto.UserId, 
            createDto.AiAgentId, 
            createDto.IsDefault);

        var responseDto = ConvertToDto(userAIAgent);
        return Ok(responseDto);
    }

    [HttpPut("{id}")]
    public async Task<ActionResult<UserAIAgentDto>> Update(Guid id, [FromBody] UpdateUserAIAgentDto updateDto)
    {
        _logger.LogDebug("Updating UserAIAgent {Id}", id);

        var userAIAgent = await _userAIAgentService.UpdateAsync(id, updateDto.IsDefault);
        if (userAIAgent == null)
        {
            return NotFound();
        }

        var responseDto = ConvertToDto(userAIAgent);
        return Ok(responseDto);
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> Delete(Guid id)
    {
        _logger.LogDebug("Deleting UserAIAgent {Id}", id);

        var deleted = await _userAIAgentService.DeleteAsync(id);
        if (!deleted)
        {
            return NotFound();
        }

        return NoContent();
    }

    [HttpGet("{id}")]
    public async Task<ActionResult<UserAIAgentDto>> GetById(Guid id)
    {
        var userAIAgent = await _userAIAgentService.GetByIdAsync(id);
        if (userAIAgent == null)
        {
            return NotFound();
        }

        var responseDto = ConvertToDto(userAIAgent);
        return Ok(responseDto);
    }

    [HttpGet("user/{userId}")]
    public async Task<ActionResult<List<UserAIAgentDto>>> GetByUserId(Guid userId)
    {
        _logger.LogDebug("Finding UserAIAgents by user ID: {UserId}", userId);

        var userAIAgents = await _userAIAgentService.GetByUserIdAsync(userId);
        var responseDtos = userAIAgents.Select(ConvertToDto).ToList();

        return Ok(responseDtos);
    }

    [HttpGet("ai-agent/{aiAgentId}")]
    public async Task<ActionResult<List<UserAIAgentDto>>> GetByAiAgentId(Guid aiAgentId)
    {
        _logger.LogDebug("Finding UserAIAgents by AI agent ID: {AiAgentId}", aiAgentId);

        var userAIAgents = await _userAIAgentService.GetByAiAgentIdAsync(aiAgentId);
        var responseDtos = userAIAgents.Select(ConvertToDto).ToList();

        return Ok(responseDtos);
    }

    [HttpGet("user/{userId}/ai-agent/{aiAgentId}")]
    public async Task<ActionResult<UserAIAgentDto>> GetByUserIdAndAiAgentId(Guid userId, Guid aiAgentId)
    {
        _logger.LogDebug("Finding UserAIAgent by user ID: {UserId} and AI agent ID: {AiAgentId}", userId, aiAgentId);

        var userAIAgent = await _userAIAgentService.GetByUserIdAndAiAgentIdAsync(userId, aiAgentId);
        if (userAIAgent == null)
        {
            return NotFound();
        }

        var responseDto = ConvertToDto(userAIAgent);
        return Ok(responseDto);
    }

    [HttpGet("default/{isDefault}")]
    public async Task<ActionResult<List<UserAIAgentDto>>> GetByIsDefault(bool isDefault)
    {
        _logger.LogDebug("Finding UserAIAgents by isDefault: {IsDefault}", isDefault);

        var userAIAgents = await _userAIAgentService.GetByIsDefaultAsync(isDefault);
        var responseDtos = userAIAgents.Select(ConvertToDto).ToList();

        return Ok(responseDtos);
    }

    [HttpGet("user/{userId}/default")]
    public async Task<ActionResult<UserAIAgentDto>> GetByUserIdAndIsDefault(Guid userId)
    {
        _logger.LogDebug("Finding default UserAIAgent for user ID: {UserId}", userId);

        var userAIAgent = await _userAIAgentService.GetByUserIdAndIsDefaultAsync(userId, true);
        if (userAIAgent == null)
        {
            return NotFound();
        }

        var responseDto = ConvertToDto(userAIAgent);
        return Ok(responseDto);
    }

    [HttpGet("user/{userId}/ai-agent/{aiAgentId}/exists")]
    public async Task<ActionResult<bool>> ExistsByUserIdAndAiAgentId(Guid userId, Guid aiAgentId)
    {
        _logger.LogDebug("Checking if UserAIAgent exists by user ID: {UserId} and AI agent ID: {AiAgentId}", userId, aiAgentId);

        var exists = await _userAIAgentService.ExistsByUserIdAndAiAgentIdAsync(userId, aiAgentId);
        return Ok(exists);
    }

    [HttpGet("count/ai-agent/{aiAgentId}")]
    public async Task<ActionResult<int>> CountByAiAgentId(Guid aiAgentId)
    {
        _logger.LogDebug("Counting UserAIAgents by AI agent ID: {AiAgentId}", aiAgentId);

        var count = await _userAIAgentService.CountByAiAgentIdAsync(aiAgentId);
        return Ok(count);
    }

    [HttpGet("count/user/{userId}")]
    public async Task<ActionResult<int>> CountByUserId(Guid userId)
    {
        _logger.LogDebug("Counting UserAIAgents by user ID: {UserId}", userId);

        var count = await _userAIAgentService.CountByUserIdAsync(userId);
        return Ok(count);
    }

    [HttpGet("count/default/{isDefault}")]
    public async Task<ActionResult<int>> CountByIsDefault(bool isDefault)
    {
        _logger.LogDebug("Counting UserAIAgents by isDefault: {IsDefault}", isDefault);

        var count = await _userAIAgentService.CountByIsDefaultAsync(isDefault);
        return Ok(count);
    }

    [HttpPut("{id}/set-default")]
    public async Task<ActionResult<UserAIAgentDto>> SetAsDefault(Guid id, [FromQuery] bool isDefault)
    {
        _logger.LogDebug("Setting UserAIAgent as default with ID: {Id} and status: {IsDefault}", id, isDefault);

        var updated = await _userAIAgentService.SetAsDefaultAsync(id, isDefault);
        if (updated == null)
        {
            return NotFound();
        }

        var responseDto = ConvertToDto(updated);
        return Ok(responseDto);
    }

    [HttpDelete("user/{userId}/clear-defaults")]
    public async Task<IActionResult> ClearDefaultForUser(Guid userId)
    {
        _logger.LogDebug("Clearing default UserAIAgent for user: {UserId}", userId);

        await _userAIAgentService.ClearDefaultForUserAsync(userId);
        return NoContent();
    }

    [HttpPost("assign")]
    public async Task<ActionResult<UserAIAgentDto>> AssignUserToAgent(
        [FromQuery] Guid userId,
        [FromQuery] Guid aiAgentId,
        [FromQuery] bool isDefault = false)
    {
        _logger.LogDebug("Assigning user: {UserId} to AI agent: {AiAgentId} with default: {IsDefault}",
            userId, aiAgentId, isDefault);

        var userAIAgent = await _userAIAgentService.CreateAsync(userId, aiAgentId, isDefault);
        var responseDto = ConvertToDto(userAIAgent);

        return Ok(responseDto);
    }

    [HttpDelete("user/{userId}/ai-agent/{aiAgentId}")]
    public async Task<IActionResult> RemoveUserFromAgent(Guid userId, Guid aiAgentId)
    {
        _logger.LogDebug("Removing user: {UserId} from AI agent: {AiAgentId}", userId, aiAgentId);

        var removed = await _userAIAgentService.RemoveUserFromAgentAsync(userId, aiAgentId);
        if (!removed)
        {
            return NotFound();
        }

        return NoContent();
    }

    private UserAIAgentDto ConvertToDto(UserAIAgent userAIAgent)
    {
        return new UserAIAgentDto
        {
            Id = userAIAgent.Id,
            CompanyId = userAIAgent.CompanyId,
            CompanyName = userAIAgent.Company?.Name,
            UserId = userAIAgent.UserId,
            UserName = userAIAgent.User?.Name,
            AiAgentId = userAIAgent.AiAgentId,
            AiAgentName = userAIAgent.AiAgent?.Name,
            IsDefault = userAIAgent.IsDefault,
            AssignedAt = userAIAgent.AssignedAt,
            CreatedAt = userAIAgent.CreatedAt,
            UpdatedAt = userAIAgent.UpdatedAt
        };
    }
}